Add timeout, maximum failed retries and resume to streams.

elnuno opened the pull request:
Fixes #778

## Feature Summary and Justification

Allows limiting the fetching loop of stream by either elapsed time since last successful fetch happened or number of failed ones. On reaching that limit, an exception is raised that stores enough data to resume the stream at a later time.

This is all experimental, so we can discuss the design.

bboe commented:
I'm excited to take a look at this work. I'll try to do so tonight (PDT), but it may not be until the weekend.

elnuno commented:
Thank you for the review!

Implemented the pause_after design. Should the counter of failed attempts be set to 0 when we yield None? (makes the most sense when writing the code, but still surprised me when testing 😄 )

I don't think the tests actually exercise the code the way they should, but not sure how to test other than get two sequential Nones (i.e., how do we know it works when a new comment does arrive?). Maybe if we record two cassettes, while adding a real new comment to a sub, we can use that to test that it goes back to yielding fresh items? Run one loop with one cassette, pause on None, resume on the other cassette... I'll test it later.

Added a lot of help in docstrings so we can also prune what's unnecessary or repetitive and better see what should be reworded or added.

Once things look good enough I'll be glad to make the same changes to submissions. And if you think it streams should be used somewhere else we could add them there.

bboe commented:
> Maybe if we record two cassettes, while adding a real new comment to a sub, we can use that to test that it goes back to yielding fresh items? Run one loop with one cassette, pause on None, resume on the other cassette... I'll test it later.

If you have it run on a test subreddit, you should be able to call the method with `pause_after` set to None. Let it run for a few seconds before adding a new comment/submission. At that point it should pick it up and the cassette is good to go. Then you can retroactively update the test case to assert that at least one `None` is yielded, or the exact number based on the cassette.

bboe commented on praw/models/reddit/subreddit.py:
```diff
@@ -1722,8 +1738,59 @@ def comments(self):
+        .. code:: python
+
+           help = reddit.subreddit('help')
+           help_stream = help.stream.comments(pause_after=100)
```
Consider using smaller values. With 1 minute delays between the latter network requests that result in no new data, a `pause_after` value of 100 will likely take over an hour before timing out. I imagine people will likely want to use this feature for approximately a 1-5 minute (pause_after of 4-10) timeout in addition to the no-delay timeout.

bboe commented on praw/models/util.py:
```diff
@@ -77,12 +77,26 @@ def permissions_string(permissions, known_permissions):
     exponential_counter = ExponentialCounter(max_counter=64)
     seen_fullnames = BoundedSet(301)
     without_before_counter = 0
+    failed_fetches = 0
```
Can we name this something other than `failed`? The request didn't fail, it simply did not have new results. How about `responses_without_new` or something to that effect?

elnuno replied:
Changed to `responses_without_new` and also stopped checking that pause_after is None on every iteration.

bboe commented on praw/models/util.py:
```diff
@@ -101,5 +115,11 @@ def stream_generator(function):
         before_fullname = newest_fullname
         if found:
             exponential_counter.reset()
+            failed_fetches = 0
```
Yes, the counter should be reset whenever it succeeds.

bboe commented on praw/models/util.py:
```diff
@@ -101,5 +115,11 @@ def stream_generator(function):
+            if pause_after is not None and failed_fetches > pause_after:
+                failed_fetches = 0
+                yield None
+                continue
```
Can you use `else` to protect the sleep, rather than use continue here?

elnuno replied:
Done!

elnuno commented on praw/models/reddit/subreddit.py:
```diff
@@ -1708,12 +1708,28 @@ def __init__(self, subreddit):
         self.subreddit = subreddit
 
-    def comments(self):
+    def comments(self, pause_after=None):
```
Done, much cleaner, thanks!

elnuno commented on praw/models/reddit/subreddit.py:
```diff
@@ -1708,12 +1708,28 @@ def __init__(self, subreddit):
         Comments are yielded oldest first. Up to 100 historical comments will
         initially be returned.
 
+        It's possible to stop the fetching loop on each failure so the calling
```
Removed and tried to make the style more consistent. Not sure whether the first example (which doesn't pass pause_after) adds values. It's simple, which is a positive.

elnuno commented on praw/models/reddit/subreddit.py:
```diff
@@ -1708,12 +1708,28 @@ def __init__(self, subreddit):
+        process comments as they are yielded (at the expense of blocking until
+        new data arrives).
+
+        :param pause_after: A integer representing the maximum times to try
```
Deal! :)

elnuno commented on praw/models/reddit/subreddit.py:
```diff
@@ -1722,8 +1738,59 @@ def comments(self):
+
+        .. code:: python
+
+           rdev = reddit.subreddit('redditdev')
```
Renamed and also com -> comment, thanks!

elnuno commented:
Thanks for the review! I'll be able to come back to this on Monday night PDT-ish.

bboe commented on tests/integration/models/reddit/test_subreddit.py:
```diff
@@ -1046,6 +1046,31 @@ def test_submissions(self):
+                                        'comments_pause'):
+            stream = self.reddit.subreddit('kakapo').stream
+            with_pause = stream.comments(pause_after=0)
+            for i in range(10):
```
This might never actually hit a pause because the first 100 items should all be comments. I think, however, it's fine to delete this test and rely just on the other one since it includes a pause.

elnuno commented:
Thanks for the review and guidance, it's great working with you!

Squashed all but two new commits (to ease review): one fixes the test, the other adds a CHANGES entry. 

Should we merge comment stream first and the work on submissions, or is it better to land both at once? Either way works for me :)

Sorry about the delay, had unforeseen stuff sneak up on me...

bboe commented:
Finishing up as https://github.com/praw-dev/praw/pull/786. Thanks!

bboe commented:
Squashed and merged as #786. Thanks a ton for the great work.

Would you be up for adding `**stream_options` to the submissions, inbox, and subreddits streams?
